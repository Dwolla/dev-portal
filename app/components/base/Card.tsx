import React from "react";
import styled from "@emotion/styled";
import Link from "next/link";
import classnames from "../../modules/classnames";
import ConditionalWrapper from "../util/ConditionalWrapper";
import Badge from "./Badge";
import { ReactComponent as NewTabIcon } from "../../../assets/images/component-icons/open-in-new-tab-icon.svg";
import {
  GREY_2,
  GREY_4,
  WHITE_PRIMARY,
  HEADLINE_TEXT,
  PARAGRAPH_TEXT,
} from "../colors";
import { BOX_SHADOW_4 } from "../shadowDepths";
import { ROBOTO, POPPINS } from "../typography";
import { breakDown } from "../breakpoints";

const CardStyle = styled.div`
  height: 100%;
  padding: 30px;
  box-sizing: border-box;
  border: 1px solid ${GREY_2};
  border-radius: 5px;
  background-color: ${WHITE_PRIMARY};
  position: relative;
  :hover {
    cursor: pointer;
    box-shadow: ${BOX_SHADOW_4};
  }

  &.center {
    max-width: none;
    height: auto;
    padding: 0 30px;
    position: relative;
    text-align: center;
    @media (${breakDown("xs")}) {
      width: 100%;
    }
  }

  &.flex {
    display: flex;
  }
`;

const StyledLink = styled.a`
  text-decoration: none;
`;

const StyledIcon = styled.img`
  height: 48px;
  width: 48px;
  &.center {
    margin-top: 24px;
    margin-left: auto;
    margin-right: auto;
  }
`;

const BadgeStyle = styled.div`
  position: absolute;
  top: 30px;
  right: 30px;
`;

const LinkIconStyle = styled.div`
  position: absolute;
  top: 8px;
  right: 8px;

  svg {
    g {
      fill: ${GREY_4};
    }
  }
`;

const StyledFlexDiv = styled.div`
  padding: 0 30px;
`;

const TopicStyle = styled.div`
  color: ${HEADLINE_TEXT};
  font-family: ${POPPINS};
  font-size: 18px;
  line-height: 26px;
  margin-top: 30px;
  margin-bottom: 15px;
  &.center {
    margin: 11px auto;
  }
  &.flex {
    margin-top: unset;
  }
`;

const DescriptionStyle = styled.div`
  color: ${PARAGRAPH_TEXT};
  font-family: ${ROBOTO};
  font-size: 15px;
  line-height: 25px;
  margin-top: 15px;
  &.center {
    margin: 11px auto 18px;
  }
`;

type Props = {
  link?: { href?: string; external?: boolean };
  centerAlign?: boolean;
  isFlex?: boolean;
  icon: string;
  badge?: string;
  topic: string;
  description: string;
};

function CoreCard({
  description,
  badge,
  centerAlign,
  isFlex,
  icon,
  link,
  topic,
}: Props) {
  return (
    <CardStyle
      className={classnames({ center: centerAlign }, { flex: isFlex })}
    >
      <StyledIcon
        src={icon}
        alt=""
        className={classnames({ center: centerAlign })}
      />

      {badge && (
        <BadgeStyle>
          <Badge text={badge} />
        </BadgeStyle>
      )}

      {link && link.external && (
        <LinkIconStyle>
          <NewTabIcon width={13} />
        </LinkIconStyle>
      )}
      <ConditionalWrapper
        condition={isFlex}
        wrapper={(children) => <StyledFlexDiv>{children}</StyledFlexDiv>}
      >
        <TopicStyle
          className={classnames({ center: centerAlign }, { flex: isFlex })}
        >
          {topic}
        </TopicStyle>

        <DescriptionStyle className={classnames({ center: centerAlign })}>
          {description}
        </DescriptionStyle>
      </ConditionalWrapper>
    </CardStyle>
  );
}

function Card({ link, ...rest }: Props) {
  return link && link.href ? (
    <Link href={link.href} passHref>
      <StyledLink
        target={link.external ? "_blank" : undefined}
        rel={link.external && "noopener noreferrer"}
      >
        <CoreCard link={link} {...rest} />
      </StyledLink>
    </Link>
  ) : (
    <CoreCard link={link} {...rest} />
  );
}

export default Card;
