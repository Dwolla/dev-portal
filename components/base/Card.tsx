import React from "react";
import styled from "@emotion/styled";

import Badge from "./Badge";
import newTabIcon from "../../public/images/open-in-new-tab-icon-light-grey.svg";
import {
  GREY_2,
  WHITE_PRIMARY,
  HEADLINE_TEXT,
  PARAGRAPH_TEXT,
} from "../colors";
import { BOX_SHADOW_4 } from "../shadowDepths";
import { ROBOTO, POPPINS } from "../typography";

type Props = {
  link?: boolean;
  centerAlign?: boolean;
  icon: string;
  badge?: string;
  topic: string;
  description: string;
};

function Card({ description, badge, centerAlign, icon, link, topic }: Props) {
  return (
    <CardStyle className={centerAlign ? "center" : ""}>
      <IconStyle className={centerAlign ? "center" : ""}>
        <img src={icon} alt="" />
      </IconStyle>

      {badge && (
        <BadgeStyle>
          <Badge text={badge} />
        </BadgeStyle>
      )}

      {link && (
        <LinkStyle>
          <img src={newTabIcon} alt="" />
        </LinkStyle>
      )}

      <TopicStyle className={centerAlign ? "center" : ""}>{topic}</TopicStyle>

      <DescriptionStyle className={centerAlign ? "center" : ""}>
        {description}
      </DescriptionStyle>
    </CardStyle>
  );
}

const CardStyle = styled.div`
  height: 305px;
  width: 300px;
  padding: 30px;
  box-sizing: border-box;
  border: 1px solid ${GREY_2};
  border-radius: 5px;
  background-color: ${WHITE_PRIMARY};
  position: relative;
  :hover {
    cursor: pointer;
    box-shadow: ${BOX_SHADOW_4};
  }
  &.center {
    height: auto;
    width: 656px;
    padding: 0px;
    position: relative;
    text-align: center;
  }
`;

const IconStyle = styled.div`
  height: 48px;
  width: 48px;
  &.center {
    margin-top: 24px;
    margin-left: auto;
    margin-right: auto;
  }
`;

const BadgeStyle = styled.div`
  position: absolute;
  top: 30px;
  right: 30px;
`;

const LinkStyle = styled.div`
  position: absolute;
  top: 8px;
  right: 8px;
`;

const TopicStyle = styled.div`
  color: ${HEADLINE_TEXT};
  font-family: ${POPPINS};
  font-size: 18px;
  line-height: 26px;
  margin-top: 30px;
  margin-bottom: 15px;
  &.center {
    margin: 11px auto;
  }
`;

const DescriptionStyle = styled.div`
  color: ${PARAGRAPH_TEXT};
  font-family: ${ROBOTO};
  font-size: 15px;
  line-height: 25px;
  margin-top: 15px;
  &.center {
    margin: 11px auto 18px;
  }
`;

export default Card;
