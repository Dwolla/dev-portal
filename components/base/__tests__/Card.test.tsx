import renderer from "react-test-renderer";
import Card from "../Card";
import icon from "../../public/images/guides-icon-large.svg";

test("default", () => {
  const tree = renderer
    .create(<Card icon={icon} topic="topic" description="Description" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Link", () => {
  const tree = renderer
    .create(<Card link icon={icon} topic="topic" description="Description" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Badge", () => {
  const tree = renderer
    .create(
      <Card
        badge="BADGETEXT"
        icon={icon}
        topic="topic"
        description="Description"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Link and Badge", () => {
  const tree = renderer
    .create(
      <Card
        link
        badge="BADGETEXT"
        icon={icon}
        topic="topic"
        description="Description"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Center Align", () => {
  const tree = renderer
    .create(
      <Card centerAlign icon={icon} topic="topic" description="Description" />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Center Align and Link", () => {
  const tree = renderer
    .create(
      <Card
        link
        centerAlign
        icon={icon}
        topic="topic"
        description="Description"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Center Align and Badge", () => {
  const tree = renderer
    .create(
      <Card
        centerAlign
        badge="BADGETEXT"
        icon={icon}
        topic="topic"
        description="Description"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("with Center Align, Link and Badge", () => {
  const tree = renderer
    .create(
      <Card
        link
        centerAlign
        badge="BADGETEXT"
        icon={icon}
        topic="topic"
        description="Description"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
