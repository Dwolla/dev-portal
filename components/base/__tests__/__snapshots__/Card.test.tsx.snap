// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<div
  className=" css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className=" css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className=" css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className=" css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Badge 1`] = `
<div
  className=" css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className=" css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-vz2qld-BadgeStyle eguc8hf2"
  >
    <div
      className="css-16zxq88-StyledBadge e1w3mjmy0"
    >
      BADGETEXT
    </div>
  </div>
  <div
    className=" css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className=" css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Center Align 1`] = `
<div
  className="center css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className="center css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="center css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className="center css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Center Align and Badge 1`] = `
<div
  className="center css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className="center css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-vz2qld-BadgeStyle eguc8hf2"
  >
    <div
      className="css-16zxq88-StyledBadge e1w3mjmy0"
    >
      BADGETEXT
    </div>
  </div>
  <div
    className="center css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className="center css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Center Align and Link 1`] = `
<div
  className="center css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className="center css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-hpm16i-LinkStyle eguc8hf3"
  >
    <svg
      width={13}
    />
  </div>
  <div
    className="center css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className="center css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Center Align, Link and Badge 1`] = `
<div
  className="center css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className="center css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-vz2qld-BadgeStyle eguc8hf2"
  >
    <div
      className="css-16zxq88-StyledBadge e1w3mjmy0"
    >
      BADGETEXT
    </div>
  </div>
  <div
    className="css-hpm16i-LinkStyle eguc8hf3"
  >
    <svg
      width={13}
    />
  </div>
  <div
    className="center css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className="center css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Link 1`] = `
<div
  className=" css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className=" css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-hpm16i-LinkStyle eguc8hf3"
  >
    <svg
      width={13}
    />
  </div>
  <div
    className=" css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className=" css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;

exports[`with Link and Badge 1`] = `
<div
  className=" css-qhvpk8-CardStyle eguc8hf0"
>
  <img
    alt=""
    className=" css-142sfdf-StyledIcon eguc8hf1"
    src="test-image-stub"
  />
  <div
    className="css-vz2qld-BadgeStyle eguc8hf2"
  >
    <div
      className="css-16zxq88-StyledBadge e1w3mjmy0"
    >
      BADGETEXT
    </div>
  </div>
  <div
    className="css-hpm16i-LinkStyle eguc8hf3"
  >
    <svg
      width={13}
    />
  </div>
  <div
    className=" css-bgc4kb-TopicStyle eguc8hf4"
  >
    topic
  </div>
  <div
    className=" css-mk6kj1-DescriptionStyle eguc8hf5"
  >
    Description
  </div>
</div>
`;
